### ISSUE TITLE
All WebSockets fail to connect with "Invalid frame header" after any deploy. Only node-red restart fixes.

### ISSUE THREAD

**Issue by Deggle** [node-red/node-red#1717]
### What are the steps to reproduce?

1. Create web socket nodes.
2. User browser to connect (e.g. http://websocket.org/echo.html)
3. Websocket seems to work, disconnect and reconnects fine.
4. Move node to enable "Deploy" button.
5. Deploy change.
6. Websocket fails to connect with console error.

I was originally running via a reverse proxy which I assumed was causing this, but I've moved to direct (just on the same LAN) connection to HTTPS/WSS (also same with HTTP/WS) and it's still happening.

From trial and error I learnt that it's fine after restarting node-red but once a deploy has happened it breaks it. I did wonder if this only happened if there was a live connection that got cut off when it deployed but tested that and it still breaks even if you restart, open admin, deploy, then try to connect.

Assume this can't be a common issue as it would be very immediately apparent to other uesrs too!?

### What happens?

**VM150:164 WebSocket connection to 'wss://xxx/yyy' failed: Invalid frame header**

### What do you expect to happen?

It to connect correctly and not display error.

### Please tell us about your environment:

- [X] Node-RED version: 0.18.4
- [X] node.js version: 8.11.1
- [X] npm version: 5.6.0
- [X] Platform/OS: Windows Server
- [X] Browser: Chrome (same issue on many others)

**Comment by knolleary** [comment]
Just to clarify, this is a websocket node connecting out to a remote server, rather than an inbound connection?

**Comment by knolleary** [comment]
Ah no, it is inbound connections. Thanks for the issue, will see if we are able to reproduce it.

**Comment by Deggle** [comment]
Just an update that the latest version (0.18.5) is still presenting this "Invalid frame header" error after deploy. It's fine once NR is restarted, but first deploy breaks it.

**Comment by knolleary** [comment]
Hi @Deggle - thanks for the test flow.

Running it locally, I cannot recreate the issue. I am able to reconnect to the websocket after a deploy.

How exactly are you running node-red? Is it in a docker image or similar? Or are you simply running the node process directly on your laptop/server..?

**Comment by Deggle** [comment]
I'm running on Windows Server and as a Windows Service. However, I've also just tried running from the console (navigate to .node-red in the user profile directory then just 'node-red') and the same happens.

So running directly and seeing the same which is very odd then. Can you think of anything I can try next? Perhaps clear out to a default/clean install and test it (is there an easy way to do this)?

**Comment by knolleary** [comment]
I wonder if the Windows Server firewall is doing something here - and it takes a restart of the NR process to free something up in the network layer.

I'm really not sure what we can do here - it isn't an issue we've seen before and we're not able to recreate. We don't have the resources to test on Windows Server so unlikely to make any progress ourselves.

**Comment by Deggle** [comment]
Tried disabling the FW completely, but not the issue. However, I've made some progress...

If I clean our all my flows, the problem goes (after a restart) and deploys don't break WS. I'm going to start dropping flows one at a time until I get to the bottom of this... hopefully I'll be back to close this off soon!

**Comment by Deggle** [comment]
**Found it!**

The culprit is the worldmap nodes (_node-red-contrib-web-worldmap_)!

If I remove all the worldmap nodes and restart I no longer have the WS issue after a deploy. Add a single worldmap node back on (even if underwired) then deploy all is ok, but if I make a change and deploy again the "_Invalid frame header_" problem returns again until restart.

I don't really use worldmap so no big loss to drop it for now.

Thanks for your help, and sorry for taking up your time with something I should have tried harder to diagnose myself first! I'll leave this open in case you want to comment, otherwise please go ahead and close.

As ever, keep up the great work!

**Comment by dceejay** [comment]
which version of worldmap are/were you using ?

**Comment by dceejay** [comment]
Your initial example doesn't work as both nodes are set to listen - I changed it to this. 

```
[{"id":"aa043446.27d078","type":"websocket in","z":"745a133b.dd6dec","name":"In Socket","server":"","client":"34530aa6.0afdfe","x":420,"y":520,"wires":[["5bca8e0.ce2c7f4"]]},{"id":"e6bd412c.059528","type":"websocket out","z":"745a133b.dd6dec","name":"Out Socket","server":"ad0079f0.a02928","client":"","x":610,"y":580,"wires":[]},{"id":"d78d6cda.52572","type":"inject","z":"745a133b.dd6dec","name":"","topic":"","payload":"","payloadType":"date","repeat":"","crontab":"","once":false,"onceDelay":0.1,"x":420,"y":580,"wires":[["e6bd412c.059528"]]},{"id":"5bca8e0.ce2c7f4","type":"debug","z":"745a133b.dd6dec","name":"","active":true,"tosidebar":true,"console":false,"tostatus":false,"complete":"false","x":610,"y":520,"wires":[]},{"id":"34530aa6.0afdfe","type":"websocket-client","z":"","path":"ws://localhost:1880/api/test","tls":"","wholemsg":"false"},{"id":"ad0079f0.a02928","type":"websocket-listener","z":"","path":"/api/test","wholemsg":"false"}]
```

**Comment by Deggle** [comment]
My test nodes both listened for an external connection (as that's how I was using WS) then one would send timestamp, and the other receive and debug. Your in/out model would make more sense for quick testing, but that doesn't even connect up for me...!?

I've just added the latest version of worldmap and am having the same issue. Drop node, deploy, move node, deploy then all WS listeners break on all flows.

**Comment by oliv3** [comment]
I also have this bug in 0.19.5.
I have flows that use the worldmap node.
Today, I was experimenting with websockets using an example from [this blog post](https://tech.scargill.net/a-node-red-websockets-web-page/). All was working fine.
Then, I made some modifications to the nodes and deployed. From this moment on, websocket connections to the test application started to fail.
Firefox was complaining with `The connection to wss://example.net/app was interrupted while the page was loading.`.
Chrome was complaining with `WebSocket connection to 'wss://example.net/app' failed: Invalid frame header`.
After some googling, I came across [this link](https://stackoverflow.com/questions/14140414/websocket-was-interrupted-while-page-is-loading-on-firefox-for-socket-io) and [this other link](https://github.com/socketio/socket.io/issues/2016).
After more googling, found this issue and can confirm the bug is still present.
To reproduce, just move a node, deploy, and it breaks. Restart the server and it's working again.
Don't know if this helps, but while the example application was broken, the worldmap was still functioning.
Hope you can find a fix soon !

**Comment by dceejay** [comment]
which browser are you using - the first post you point to indicates a bug in firefox.

**Comment by oliv3** [comment]
* Firefox 60.5.1esr (64-bit)
* Chromium 72.0.3626.122

**Comment by oliv3** [comment]
I don't think it's a browser issue. The code in the blog post was:
```html
<body onload="wsConnectC();" onunload="ws.disconnect;">
```
I checked the source code from worldmap.html:
```html
<body onunload="ws.send(JSON.stringify({'action':'end'}));">
```
so they are doing quite the same thing.
Best way to reproduce IMHO would be to run the example from the mentioned blog post along with a flow using the worldmap node ?

**Comment by dceejay** [comment]
you could try that fix - edit the worldmap/index.html file line you found to be

    <body onunload="ws.send(JSON.stringify({'action':'end'})); ws.disconnect;">

and edit the `worldmap.appcache` slightly (date or version number) - then reload the browser a couple of times to ensure the new version is in the cache... let me know what happens

**Comment by oliv3** [comment]
Unfortunately this doesn't work... Here's what I changed:
* in worldmap/index.html:
```diff
- <body onunload="ws.send(JSON.stringify({'action':'end'}));">
+ <body onunload="ws.send(JSON.stringify({'action':'end'})); ws.disconnect;">
```
* in worldmap/worldmap.appcache:
```diff
- # date: Feb 23rd 2019 - v1.5.26
+ # date: Feb 23rd 2019 - v1.5.26.0
```
Then I force reloaded the worldmap page 3/5 times, moved some nodes and deployed, it broke again. FWIW Node-RED is running in a docker container, so I stopped docker-compose and restarted, everything worked again. Retried the procedure, same result.

**Comment by dceejay** [comment]
Can you share your test flow ? Thanks

**Comment by oliv3** [comment]
Sure, here it is:
```json
[
    {
        "id": "d3c0fad2.abd7c",
        "type": "tab",
        "label": "WebSockets test",
        "disabled": false,
        "info": ""
    },
    {
        "id": "5b1b0abf.d7e2ac",
        "type": "change",
        "z": "d3c0fad2.abd7c",
        "name": "",
        "rules": [
            {
                "t": "delete",
                "p": "_session",
                "pt": "msg"
            }
        ],
        "action": "",
        "property": "",
        "from": "",
        "to": "",
        "reg": false,
        "x": 440,
        "y": 100,
        "wires": [
            [
                "111c4586.bed40a"
            ]
        ]
    },
    {
        "id": "46f5783d.679848",
        "type": "http in",
        "z": "d3c0fad2.abd7c",
        "name": "",
        "url": "/csuapp",
        "method": "get",
        "upload": false,
        "swaggerDoc": "",
        "x": 190,
        "y": 240,
        "wires": [
            [
                "6c6168c8.2ef968",
                "856888d3.b5637"
            ]
        ]
    },
    {
        "id": "170913f2.abe3ac",
        "type": "http response",
        "z": "d3c0fad2.abd7c",
        "name": "csu app response",
        "statusCode": "",
        "headers": {
            "content-type": "text/html"
        },
        "x": 650,
        "y": 240,
        "wires": []
    },
    {
        "id": "6c6168c8.2ef968",
        "type": "debug",
        "z": "d3c0fad2.abd7c",
        "name": "/csu query",
        "active": true,
        "tosidebar": true,
        "console": false,
        "tostatus": false,
        "complete": "payload",
        "x": 410,
        "y": 160,
        "wires": []
    },
    {
        "id": "856888d3.b5637",
        "type": "template",
        "z": "d3c0fad2.abd7c",
        "name": "",
        "field": "payload",
        "fieldType": "msg",
        "format": "handlebars",
        "syntax": "plain",
        "template": "<!DOCTYPE HTML>\n<html>\n\n<head>\n<title>Simple Display</title>\n<meta name=\"viewport\" content=\"width=device-width, initial-scale=1\">\n<link rel=\"stylesheet\" href=\"https://ajax.googleapis.com/ajax/libs/jquerymobile/1.4.5/jquery.mobile.min.css\">\n<script src=\"https://ajax.googleapis.com/ajax/libs/jquery/2.1.3/jquery.min.js\"></script>\n<script src=\"https://ajax.googleapis.com/ajax/libs/jquerymobile/1.4.5/jquery.mobile.min.js\"></script>\n</head>\n\n<script type=\"text/javascript\">\n\nvar server = window.location.hostname;\nvar topics = {};\nvar wsUriC = \"wss://\"+server+\"/csu\";\nvar ws;\nfunction wsConnectC() {\nconsole.log(\"connect\",wsUriC);\nws = new WebSocket(wsUriC);\nws.onmessage = function(msg) {\nvar line = \"\";\nconsole.log(msg.data);\nvar fromPage=JSON.parse(msg.data);\nif (fromPage.part1!=\"\") $(\"#part1\").html(fromPage.part1); else $(\"#part1\").html(\"!Empty!\");\n$(\"#part2\").html(fromPage.part2+\"c\");\n}\nws.onopen = function() {\n$(\"#status\").html(\"connected\");\nconsole.log(\"connected\");\n}\nws.onclose = function() {\n$(\"#status\").html(\"not connected\");\nsetTimeout(wsConnectC,1000);\n}\n}\nfunction sendMessage(){\n// send message back to page in simple JSON format\n// example {“part1”:”Hello”,”part2”:”50”}\nvar toPage='{\"part1\":\"'+$(\"#txtMsg_1\").val()+'\",\"part2\":\"'+$(\"#slider_1\").val()+'\",\"part3\":\"3\"}';\nws.send(toPage);\n} // end sendMessage\n\n</script>\n\n<body onload=\"wsConnectC();\" onunload=\"ws.disconnect;\">\n<div data-role=\"page\" id=\"one\">\n<div data-role=\"header\">\n<h1>Websockets Test Page</h1>\n</div>\n<div role=\"main\" class=\"ui-content\">\n<h1>Temperature Display</h1>\n<div id=\"status\">status unknown</div>\n<input id=\"txtMsg_1\" />\n<div id=\"part1\">!Empty!</div>\n<hr/>\n<div id=\"part2\">0</div>\n<hr/>\n<label for=\"slider_1\">Input slider:</label>\n<input type=\"range\" id=\"slider_1\" value=\"60\" min=\"0\" max=\"100\"  />\n<input type=\"button\" value=\"Update\" onClick=\"sendMessage()\" />\n</div>\n</div>\n</body>\n\n</html>\n",
        "output": "str",
        "x": 400,
        "y": 240,
        "wires": [
            [
                "170913f2.abe3ac"
            ]
        ]
    },
    {
        "id": "552e9cf0.c65ba4",
        "type": "debug",
        "z": "d3c0fad2.abd7c",
        "name": "/csu websocket",
        "active": true,
        "tosidebar": true,
        "console": false,
        "tostatus": false,
        "complete": "payload",
        "x": 420,
        "y": 40,
        "wires": []
    },
    {
        "id": "887c899.f2481f8",
        "type": "websocket in",
        "z": "d3c0fad2.abd7c",
        "name": "",
        "server": "91bb9235.b5d218",
        "client": "",
        "x": 180,
        "y": 100,
        "wires": [
            [
                "5b1b0abf.d7e2ac",
                "552e9cf0.c65ba4"
            ]
        ]
    },
    {
        "id": "111c4586.bed40a",
        "type": "websocket out",
        "z": "d3c0fad2.abd7c",
        "name": "",
        "server": "91bb9235.b5d218",
        "client": "",
        "x": 680,
        "y": 100,
        "wires": []
    },
    {
        "id": "91bb9235.b5d218",
        "type": "websocket-listener",
        "z": "",
        "path": "/csu",
        "wholemsg": "false"
    }
]
```

**Comment by oliv3** [comment]
Other flow has a `worldmap` and a `worldmap in`

**Comment by dceejay** [comment]
In your template line 16 - if I change it to \

    var wsUriC = "ws://"+server+":1880/csu";

It works (I think)... .certainly you need to use port 1880 as that is where Node-RED is running. You may find wss works but may depend on your local setup.

**Comment by oliv3** [comment]
I use `wss://` because Node-RED stands behind a Nginx reverse-proxy doing SSL termination.

**Comment by oliv3** [comment]
Changing to `ws://` results in a Content-Security-Policy violation:
`Content Security Policy: The page’s settings blocked the loading of a resource at ws://example.net/csu (“default-src”).`
Wondering now why the `worldmap` node doesn't cause any problem, since the CSP looks like:
`Content-Security-Policy "default-src https: 'self'; script-src 'self' 'unsafe-inline' 'unsafe-eval' ajax.googleapis.com; style-src 'self' 'unsafe-inline' ajax.googleapis.com; img-src https: 'self' data: *.tile.openstreetmap.org server.arcgisonline.com geo.nls.uk";`

**Comment by dceejay** [comment]
well it's working without nginx... so... not sure what to suggest next

**Comment by oliv3** [comment]
I disabled all the flows containing `worldmap` nodes, now I can move nodes in my flow, deploy and it's working, still with Nginx upstream.

Questions are, then:
* why does `worldmap` alone work, even with this Nginx configuration ?
* why does it crash when there are other flows using `websocket` nodes ?

Not an expert in that domain, but couldn't it be some form of multiplexing issue ? That is, everything works fine as long as you have only **one** node using WebSockets ?

**Comment by dceejay** [comment]
could be. Worldmap uses sock.js and core Node-RED uses pure websockets. But they are going to different endpoints so should be independent.

**Comment by oliv3** [comment]
From [this comment](https://github.com/node-red/node-red/issues/1717#issuecomment-396058692) culprit would be the `worldmap` node.
Which is very unfortunate, as I need it along with (several) upcoming `websocket` nodes, and requirements here are that Node-RED has to stand behind a secure proxy...
So this is a show stopper..
So I will deploy that same test application on two different websocket end points, without worldmap, and see how it goes.

**Comment by dceejay** [comment]
Not sure if you spotted. I pushed an update to worldmap ( with one of your other requests fixed) and also included an updated socksjs library.

**Comment by oliv3** [comment]
Oh right, in https://github.com/dceejay/RedMap/issues/73 ? Thanks, will try and let you know !

**Comment by oliv3** [comment]
Just tried and it seems to fix this issue ! Will make more thorough tests tomorrow to confirm.
Thanks again !

**Comment by knolleary** [comment]
Closing as part of a tidy-up of the issue tracker — this has been inactive for a long time and much has changed in Node-RED since. If it's still relevant on a current version, please open a fresh issue with up-to-date details and we'll take a look. Thanks for raising it.
